volpiano_display_utilities/text_volpiano_alignment.py

```diff
@@ -6,238 +6,317 @@
+from .volpiano_syllabification import (
+    syllabify_volpiano,
+    preprocess_volpiano,
+    adjust_music_spacing,
+    adjust_missing_music_spacing,
 )
 
-# Matches any material before the clef, the clef itself, and
-# any following spacing in a volpiano string.
-STARTING_MATERIAL_REGEX = re.compile(r".*1-*")
-
-# Split sections of volpiano on clefs, barline markers, and missing
-# music indicators. Includes spacing (hyphens) and section markers (7's)
-# in the split, if present.
-VOLPIANO_SECTIONING_REGEX = re.compile(r"([134][\-7]*|6[\-7]*6[\-7]*)")
-
-VOLPIANO_WORD_REGEX = re.compile(r".*?---")
-
-VOLPIANO_SYLLABLE_REGEX = re.compile(r".*?-{2,3}")
+T = TypeVar("T", str, List, SyllabifiedTextSection, SyllabifiedVolpianoSection)
 
 
-def _preprocess_volpiano(raw_volpiano_str: str) -> str:
+def _zip_and_align(
+    text: List[T], volpiano: List[T], pad_text: T, pad_volpiano: T
+) -> List[Tuple[T, T]]:
     """
-    Prepares volpiano string for alignment with text:
-    - Checks for any invalid characters
-    - Ensure volpiano begins with a clef followed by three hyphens. Assume
-        that anything entered before a clef should be removed.
-    - Ensures proper spacing around barlines ("3" and "4") and missing
-        music indicators ("6")
-    - Ensures volpiano string has an ending barline ("3" or "4")
-    - Finds line and page breaks and removes them from the volpiano string
-        (they are added back in postprocessing)
-
-    raw_volpiano_str [str]: unprocessed volpiano string
-
-    returns [str]: preprocessed volpiano string
-    """
-    # Remove existing clef and any material preceding the
-    # clef. Re-add a "clean" starting clef.
-    vol_clef_rmvd = STARTING_MATERIAL_REGEX.sub("", raw_volpiano_str)
-    vol_clef_added = "1---" + vol_clef_rmvd
-    # Remove existing ending bar line and add a "clean" ending barline.
-    last_char = vol_clef_added[-1]
-    if last_char not in "34":
-        last_char = "3"
-    vol_clef_fin_bar_added = (
-        vol_clef_added.rstrip("3").rstrip("4").rstrip("-") + "---" + last_char
-    )
-    processed_vol = INVALID_VOLPIANO_CHARS_REGEX.sub("", vol_clef_fin_bar_added)
-    logging.debug("Preprocessed volpiano string: %s", processed_vol)
-    return processed_vol
+    Aligns lists of text and volpiano together and adds padding
+    if necessary. Can be used to align sections (in which case the
+    text and volpiano arguments are lists of lists) or words (in which
+    case the text and volpiano arguments are lists of strings).
 
+    text [list[T]]: text list
+    volpiano [list[T]]: volpiano list
+    pad_text [T]: value to fill text with if text is shorter than volpiano
+    pad_volpiano [T]: value to fill volpiano with if volpiano is shorter than text
 
-def _postprocess_spacing(
-    comb_text_and_vol: List[Tuple[str, str]],
-) -> List[Tuple[str, str]]:
+    returns [list[tuple[T, T]]]: zipped list with padding
```

Comment:
Using "T" only works if all the arguments are of the same type, doesn't it? If `text` is a list of `SyllabifiedTextSection`s and `volpiano` is a list of `SyllabifiedVolpianoSection`s, then what's contained in the tuple - a pair of TextSections or a pair of VolpianoSections?

I'm going to set this aside for now, since this might all be me misundersanding how `TypeVar`s work, but unless there's a demonstrable need for this function to handle a bunch of different types as arguments, I think that allowing non-`str` inputs makes it too complicated.

Reply:
I've updated the docstring with the new TypeVar definition in [61cd09e](https://github.com/DDMAL/volpiano-display-utilities/pull/26/commits/61cd09eb02bee74396652d03f319a14679a0a812).

Does that clarify? Take a look and comment or resolve as appropriate.